SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityScheduler.java

```diff
@@ -214,66 +217,81 @@ public void checkForDecomissions(SingularitySchedulerStateCache stateCache) {
 
-      final SingularityDeployStatistics deployStatistics = getDeployStatistics(pendingRequest.getRequestId(), pendingRequest.getDeployId());
+        if (!shouldScheduleTasks(updatedRequest, pendingRequest, maybePendingDeploy, maybeRequestDeployState)) {
+          LOG.debug("Pending request {} was obsolete (request {})", pendingRequest, SingularityRequestWithState.getRequestState(maybeRequest));
+          obsoleteRequests++;
+          requestManager.deletePendingRequest(pendingRequest);
+          continue;
+        }
 
-      final RequestState requestState = checkCooldown(maybeRequest.get().getState(), updatedRequest, deployStatistics);
+        int missingInstances = getNumMissingInstances(matchingTaskIds, updatedRequest, pendingRequest, maybePendingDeploy);
+        if (missingInstances == 0 && !matchingTaskIds.isEmpty() && updatedRequest.isScheduled() && pendingRequest.getPendingType() == PendingType.NEW_DEPLOY) {
+          LOG.trace("Holding pending request {} because it is scheduled and has an active task", pendingRequest);
+          heldForScheduledActiveTask++;
+          continue;
+        }
 
-      int numScheduledTasks = scheduleTasks(stateCache, updatedRequest, requestState, deployStatistics, pendingRequest, matchingTaskIds, maybePendingDeploy);
+        if (effectivePendingRequests.isEmpty()) {
+          effectivePendingRequests.add(pendingRequest);
+        } else if (pendingRequest.getPendingType() == PendingType.IMMEDIATE) {
+          effectivePendingRequests.add(pendingRequest);
+        } else if (pendingRequest.getPendingType() == PendingType.ONEOFF) {
+          effectivePendingRequests.add(pendingRequest);
+        } else {
+          // Any other subsequent requests are not honored until after the pending queue is cleared.
+        }
 
-      if (numScheduledTasks == 0 && !matchingTaskIds.isEmpty() && updatedRequest.isScheduled() && pendingRequest.getPendingType() == PendingType.NEW_DEPLOY) {
-        LOG.trace("Holding pending request {} because it is scheduled and has an active task", pendingRequest);
-        heldForScheduledActiveTask++;
-        continue;
+        requestManager.deletePendingRequest(pendingRequest);
       }
 
-      LOG.debug("Pending request {} resulted in {} new scheduled tasks", pendingRequest, numScheduledTasks);
-
-      totalNewScheduledTasks += numScheduledTasks;
+      int scheduledInstances = 0;
+      for (SingularityPendingRequest effectivePendingRequest : effectivePendingRequests) {
+        RequestState requestState = checkCooldown(maybeRequest.get().getState(), maybeRequest.get().getRequest(), deployStatistics);
+        scheduledInstances += scheduleTasks(stateCache, maybeRequest.get().getRequest(), requestState,
+            deployStatistics, effectivePendingRequest, matchingTaskIds, maybePendingDeploy);
+      }
 
-      requestManager.deletePendingRequest(pendingRequest);
+      totalNewScheduledTasks += scheduledInstances;
```

Comment:
let's keep the variable as `numScheduledTasks` since we call it tasks not instances everywhere else in this class